### PR TITLE
feat(trace): Add gcp trace fields to a logger

### PR DESCRIPTION
**Pull request by andypwarren**
This is an attempt to link logs to tracing. We currently have the ability to link logs to traces if the log has been created by our LogUnaryInterceptor or LogStreamInterceptor with the traceField config. This goes a step further and allows us to add span context to a log from a new span. 

First up the OtelProvider needs to be registered with the GCPTraceLogger e.g.
```go
traceProvider, err := otelkit.NewOtelProvider(
	...
	otelkit.WithGCPTraceLogger("gcp-project-id"),
)
```

Then you grab the log decorated with the span/trace fields when you start the span e.g.
```go
ctx, span := s.tracer.Start(stream.Context(), "UpdateRedirects")
defer span.End()
log := zerolog.Ctx(ctx)
```

### CODE REVIEW

File: tracing/otel/helpers.go [hunk 1]
```diff
@@ -21,3 +25,36 @@ func SpanRecordError(span trace.Span, err error, description string, eventOption
...
+
+// LogFromCtx returns a log from the provided context. It adds GCP trace and span fields so the log can be associated with cloud tracing
+func (tl *gcpTraceLog) LogFromCtx(ctx context.Context) *zerolog.Logger {
+	log := zerolog.Ctx(ctx)
+	span := trace.SpanFromContext(ctx)
+	fields := map[string]interface{}{}
+	if span.SpanContext().HasSpanID() {
+		fields[tl.spanFieldName] = span.SpanContext().SpanID()
+	}
+	if span.SpanContext().HasTraceID() {
```
Comment (BradleyChatha) [inline tracing/otel/helpers.go:48]: Potentially should add `&& tl.gcpProjectID != ""`

**Comment by BradleyChatha** [comment]
Looks fine to me, but a small discussion point:

I know pretty much all of our code uses zerolog, but I think we had aspirations for looking into the standard `slog` package in the future? 

Would it bite us if the generically-name/default-sounding functions & workflow like `LogFromCtx` are tied to zerolog? 

I don't have much in the way of a cleaner solution though that seems to be worth the effort, but thought whether it's something to consider/has been considered already.

**Comment by BradleyChatha** [comment]
I wonder as well what we should do with grpckit's log interceptors, since their tracing stuff is obsolete & slightly in conflict here. Topic for another time/MR though.

File: tracing/otel/helpers.go [hunk 2]
```diff
@@ -21,3 +25,36 @@ func SpanRecordError(span trace.Span, err error, description string, eventOption
...
+}
+
+type gcpTraceLog struct {
+	traceFieldName string
+	spanFieldName  string
+	gcpProjectID   string
+}
+
+// LogFromCtx returns a log from the provided context. It adds GCP trace and span fields so the log can be associated with cloud tracing
+func (tl *gcpTraceLog) LogFromCtx(ctx context.Context) *zerolog.Logger {
```
Comment (BradleyChatha) [inline tracing/otel/helpers.go:41]: Could also add a test for this function.

File: tracing/otel/provider.go [hunk 3]
```diff
@@ -183,7 +197,10 @@ func (o *OtelProvider) SetupGlobalState(ctx context.Context) (CleanupFunc, error
 //
 // If you're interested in further details, please see the `Tracer.Start` function from the OTEL SDK.
 func (o *OtelProvider) Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
-	return o.tracerFromContext(ctx).Start(ctx, spanName, opts...)
+	ctx, span := o.tracerFromContext(ctx).Start(ctx, spanName, opts...)
+	log := o.getTraceLogger.LogFromCtx(ctx)
+	ctx = log.WithContext(ctx)
```
Comment (BradleyChatha) [inline tracing/otel/provider.go:202]: Technically should add a test to ensure that the context is definitely having the logger injected, but I'll leave it up to your judgement.

**Review by BradleyChatha — CHANGES_REQUESTED** [review]

File: tracing/otel/helpers.go (hunk 1 above)
Comment (andypwarren) [reply tracing/otel/helpers.go:48]: Yeh fair enough, I'll add that

**Comment by andypwarren** [comment]
> Looks fine to me, but a small discussion point:
> 
> I know pretty much all of our code uses zerolog, but I think we had aspirations for looking into the standard `slog` package in the future?
> 
> Would it bite us if the generically-name/default-sounding functions & workflow like `LogFromCtx` are tied to zerolog?
> 
> I don't have much in the way of a cleaner solution though that seems to be worth the effort, but thought whether it's something to consider/has been considered already.

I'm happy with where we are with zerolog at the moment. I'm not seeing any benefits moving to slog at the moment aside from the fact it's the new way todo structured logging

File: tracing/otel/helpers.go (hunk 2 above)
Comment (andypwarren) [reply tracing/otel/helpers.go:41]: Yeh I wanted to but got frustrated trying to mock the span. I'll give it another stab

Comment (BradleyChatha) [reply tracing/otel/helpers.go:41]: Seems the mock tracer lets you specify some hard coded values for the span: https://github.com/open-telemetry/opentelemetry-go/blob/main/bridge/opentracing/internal/mock.go#L46

**Comment by andypwarren** [comment]
> I wonder as well what we should do with grpckit's log interceptors, since their tracing stuff is obsolete & slightly in conflict here. Topic for another time/MR though.
> 
> My initial thought is to deprecate uage of those grpckit interceptors and add them into otelkit instead, with more direct integration with otel spans.

Yes I guess they are becoming more obsolete. They are useful if we want to corrolate logs to traces if we haven't manually created the span via the trace provider and are just using the trace interceptors.

File: tracing/otel/helpers.go (hunk 2 above)
Comment (andypwarren) [reply tracing/otel/helpers.go:41]: Just ended up mocking my own span which was simple enough

**Comment by BradleyChatha** [comment]
Awesome, I know that probably felt a bit of a needless slog, but think of it as a chill end of week task 😆

**Review by BradleyChatha — APPROVED** [review]

**Comment by andypwarren** [comment]
> Awesome, I know that probably felt a bit of a needless slog, but think of it as a chill end of week task 😆

All good, I should have written the tests in the first place

**Pull request merged**